### DISCUSSION TITLE
Trigger vs DAG serialization

### DISCUSSION THREAD

**Discussion by llamageddon83** [apache/airflow#24891] [Q&A] [answered]
I have a custom operator, trigger, and hook. The folder structure is something like this:

```
|_ dags
   |__ hooks
     |__ hook.py
   |__ triggers
     |__ trigger.py
   |__ operators
     |__ operator.py
   |__ util
   dag_file.py
```

In the docker-compose they are mounted at:

```
volumes:
  - ./dags:/opt/airflow/dags
 ```

The problem I am running into is that every time I make a change in the hook or trigger file, they are not reflected unless the docker-compose is brought down. The dag changes reflect as expected. DAG fetch and update intervals are set to 60 seconds in the docker-compose.  

I have double-checked the docker container and the files in the container are up to date. There is nothing wrong with the path or anything. I have also forced airflow to reserialize (`airflow dags reserialize`). 

This is a very peculiar behavior. It seems like anything that runs on the triggerer container/pod is not updated unless the container is restarted. Am I missing something completely here or is this expected?

Thank you

**Comment by potiuk** [comment, marked as answer]
Only dags are re-imported automatically. If you have custom trigger you need to restart triggerer indeed. Custom triggers are supposed to be deployed to airflow rarely. You might want to make a PR to improve that if you want to reload triggers on the fly but this is not currently done or planned.  But contributing a PR to make it happen is most welcome.

**Reply by llamageddon83** [reply]
Thank you for the quick response.

> Custom triggers are supposed to be deployed to airflow rarely

I don't fully agree with that statement. We have a need for async operations and the support for those out of the box is minimal. There are only two async triggers available out of the box I believe (DateTimeSensorAsync and another one). 

The problem right now is that reloading a custom trigger would require a cluster restart at the minimum. And the CI/CD pipeline is set up to only copy the dags folder onto the mounted volume in the kube cluster.

**Reply by potiuk** [reply]
Again - feel free to PR it. Airflow has > 2100 contributors and mostly they are users like you who have a need to improve something. I will be happy to review such contribution. Sometimes people "expect" things to happen in the absolutely free software they get, yet they have a fantastic opportunity to pay back and contribute a feature they find useful for themselves to be also available to others.

You (or maybe someone paid by your company if your company benefits from free Airflow) are most welcome to submit such feature.

**Reply by potiuk** [reply]
Don't miss the opportunity @throwawayaccount11111 - it's a good way to become contributor to open-source and pay back for the free software !

**Reply by llamageddon83** [reply]
@potiuk Fair enough. I did start looking around where those changes would need to be made. I was just wondering if there is an existing pattern/best practice to push/promote trigger related changes without needing to redeploy airflow.

**Reply by potiuk** [reply]
As explained - you just need to restart triggerrer after changing your triggerer code. That's it.
And you could do it without redeploying airflow - but this is totally outside of realm of airflow how you restart your components. For example in kubernetes and our chart it is enough to kill triggerer pod and K8S will restart it - as long as your Triggers are refreshed (for example in shared volume or git sync).

**Reply by llamageddon83** [reply]
Got it. Thank you for the help
